File: 575/distribute-candies.go
func distributeCandies(candies []int) int {
	length := len(candies)
	if length&1 != 0 {
		return 0
	}
	half := length >> 1

	hash := make(map[int]int)

	for _, v := range candies {
		_, ok := hash[v]
		if !ok {
			hash[v] = 1
		} else {
			hash[v]++
		}
	}

	kinds := len(hash)
	if kinds >= half {
		return half
	} else {
		return kinds
	}
}

File: 575/distribute-candies.js
/**
 * @param {number[]} candies
 * @return {number}
 */
function distributeCandies(candies) {
    if (candies & 1)
        return 0;
    let kind = [...new Set(candies)].length;
    let half = candies.length >> 1;
    if (kind >= half) {
        return half;
    } else {
        return kind;
    }
}